src/paths/steiner.cpp

```diff
@@ -0,0 +1,650 @@
+ *              take only in-neighbours
+ *         \endclist
+ *
+ * \param weights The edge weights. All edge weights must be
+ *       non-negative. Additionally, no edge weight may be NaN.
+ *
+ * \return Error code:IGRAPH_ERROR
+ *
+ * Time complexity: O( 3^k ∗ V + 2^k ∗V^2 + V∗(V+E) ∗ log(V) )
+ * where V is vertices and E is edges in the graph and k is set of steiner terminals.
+ * It's recommended that V <= 50 and k < 11
+ *
+ * \sa \ref igraph_distances_johnson(), generateSubsets(), fetchIndexofMapofSets(), generate_steiner_tree_appx()
+ */
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph, const igraph_vector_int_t *steiner_terminals,
+											  const igraph_vector_t *weights, igraph_real_t *res, igraph_vector_int_t *res_tree)
+{
+	igraph_bool_t IsConnected;
+	igraph_is_connected(graph,&IsConnected,IGRAPH_WEAK);
+
+	if (!IsConnected)
+	{
+		*res = 0.0;
+		res_tree = nullptr;
+		IGRAPH_ERROR("The graph is disconnected.",IGRAPH_EINVAL);
+	}
+	igraph_integer_t no_of_vertices = igraph_vcount(graph);
+	igraph_integer_t no_of_edges = igraph_ecount(graph);
+
+	if (no_of_vertices == 0 || (no_of_vertices == 1)) // graph is empty
+	{
+		igraph_error_t x = igraph_minimum_spanning_tree(graph, res_tree,weights);
+		*res = 0.0;
+		return x;
+	}
+
+	/*
+	 *	If the steiner terminals is number of vertices in graph then problem
+	 *	is reduced to minimum spanning tree which is tractable.
+	 */
+	igraph_real_t min = igraph_vector_min(weights);
+    if (min < 0) {
+		*res = IGRAPH_INFINITY;
+        IGRAPH_ERRORF("Weight vector must be non-negative, got %g.", IGRAPH_EINVAL, min);
+	}
+	else if (min == 0){
+		*res = IGRAPH_INFINITY;
+		IGRAPH_ERROR("Weight vector contains zero weight.", IGRAPH_EINVAL);
+	}
+	if (igraph_vector_int_size(steiner_terminals) == no_of_vertices)
+	{
+
+		igraph_error_t result  = igraph_minimum_spanning_tree(graph, res_tree, weights);
+		igraph_real_t size_value = 0.0;
+		for (igraph_integer_t i = 0; i < igraph_vector_int_size(res_tree); i++)
+		{
+			size_value  += VECTOR(*weights)[VECTOR(*res_tree)[i]];
+		}
+		*res = size_value;
+		return result;
+	}
+	igraph_vector_int_t steiner_terminals_copy;
+	igraph_matrix_t dp_cache; // dynamic programming table
+	igraph_integer_t q;
+	std::set<std::set<igraph_integer_t>> allSubsets;
+	igraph_matrix_t distance;
+
+	if (igraph_vector_size(weights) != no_of_edges)
+	{
+		*res = 0;
+		IGRAPH_ERRORF("Weight vector length does not match %" IGRAPH_PRId "vec size and %" IGRAPH_PRId "edges.", IGRAPH_FAILURE, igraph_vector_size(weights), no_of_edges);
+	}
+	IGRAPH_CHECK(igraph_matrix_init(&distance, no_of_vertices, no_of_vertices));
+	IGRAPH_FINALLY(igraph_matrix_destroy, &distance);
+
+	/*
+	 *	Johnson's algorithm calculates all pairs shortest path
+	 *	and returns distance matrix. The Dreyfus - Wagner algorithm needs complete graph
+	 *   hence this step is necessary.
+	 */
+	igraph_distances_johnson(graph, &distance, igraph_vss_all(), igraph_vss_all(), weights);
```

Comment:
Make sure that all functions that return error codes are wrapped in `IGRAPH_CHECK()`, both on this line and elsewhere.